File: weaverlib/src/main/java/com/antourage/weaverlib/other/networking/Resource.java
package com.antourage.weaverlib.other.networking;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

//a generic class that describes a data with a state
public class Resource<T> {
    @NonNull
    public final State state;
    @Nullable
    public final T data;
    @Nullable
    public final String message;
    @Nullable
    public final Integer statusCode;

    private Resource(@NonNull State state, @Nullable T data, @Nullable String message, @Nullable Integer statusCode) {
        this.state = state;
        this.data = data;
        this.message = message;
        this.statusCode = statusCode;
    }

    public static <T> Resource<T> success(@NonNull T data) {
        return new Resource<>(State.SUCCESS, data, null, null);
    }

    public static <T> Resource<T> error(String msg, @Nullable T data,@Nullable Integer statusCode) {
        return new Resource<>(State.FAILURE, data, msg, statusCode);
    }

    public static <T> Resource<T> loading(@Nullable T data) {
        return new Resource<>(State.LOADING, data, null, null);
    }
}

File: ant-viewver/src/main/java/com/antourage/weaverlib/other/networking/Resource.kt
package com.antourage.weaverlib.other.networking

/**
 * A generic class that holds a value with its loading status.
 * @param <T>
</T> */
internal data class Resource<T>(val status: Status<T>) {
    companion object {
        fun <T> success(data: T?): Resource<T> {
            return Resource(
                Status.Success(
                    data
                )
            )
        }

        fun <T> cachedData(data: T?): Resource<T> {
            return Resource(
                Status.CachedData(
                    data
                )
            )
        }

        fun <T> failure(msg: String, errorCode: Int? = null): Resource<T> {
            return Resource(
                Status.Failure(
                    msg,
                    errorCode
                )
            )
        }

        fun <T> loading(): Resource<T> {
            return Resource(Status.Loading())
        }
    }
}
